pkg/base1/test-import-json.ts

```diff
@@ -0,0 +1,264 @@
+    assert.deepEqual(
+        console_errors,
+        [
+            'JSON validation error for [2]: Not a number: "c"',
+            'JSON validation error for : Not an array: "foo"'
+        ],
+        "console errors"
+    );
+});
+
+QUnit.test("import_record", function(assert) {
+    assert.deepEqual(
+        import_record({ a: "a", b: "b", c: "c" }, import_string),
+        { a: "a", b: "b", c: "c" },
+        "record of strings");
+    assert.deepEqual(
+        import_record({ a: 1, b: 2, c: 3 }, import_number),
+        { a: 1, b: 2, c: 3 },
+        "record of numbers");
+    assert.deepEqual(
+        import_record({ a: 1, b: 2, c: "c" }, import_number),
+        { a: 1, b: 2 },
+        "record of numbers with a string");
+    assert.throws(() => import_record("foo", import_string), "not a record");
+    assert.deepEqual(
+        console_errors,
+        [
+            'JSON validation error for .c: Not a number: "c"',
+            'JSON validation error for : Not an object: "foo"'
+        ],
+        "console errors"
+    );
+});
+
+QUnit.test("validate", function(assert) {
+    assert.equal(validate("test input", "foo", import_string, ""), "foo", "string");
+    assert.equal(validate("test input", 12, import_string, ""), "", "not a string");
+    assert.deepEqual(
+        console_errors,
+        [
+            'JSON validation error for test input: Not a string: 12'
+        ],
+        "console errors"
+    );
+});
+
+interface Player {
+    name: string;
+    age?: number;
+}
+
+function import_Player(val: JsonValue): Player {
+    const obj = import_json_object(val);
+    const res: Player = {
+        name: import_mandatory(obj, "name", import_string),
+    };
+    import_optional(res, obj, "age", import_number);
+    return res;
+}
+
+interface Team {
+    name: string;
+    players: Player[];
+}
+
+function import_Team(val: JsonValue): Team {
+    const obj = import_json_object(val);
+    const res: Team = {
+        name: import_mandatory(obj, "name", import_string),
+        players: import_mandatory(obj, "players", v => import_array(v, import_Player))
+    };
+    return res;
+}
+
+interface Teams {
+    [name: string]: Team;
+}
+
+function import_Teams(val: JsonValue): Teams {
+    return import_record(val, import_Team);
+}
+
+QUnit.test("objects", function(assert) {
+    const checks: [JsonValue, Teams][] = [
+        [
+            {
+                MAC: {
+                    name: "ManCity",
+                    players: [
+                        { name: "Haaland", age: 24 },
+                        { name: "De Bruyne", age: 33 }
+                    ],
+                    stadium: "City of Manchester Stadium"
+                },
+            },
+            {
+                MAC: {
+                    name: "ManCity",
+                    players: [
+                        { name: "Haaland", age: 24 },
+                        { name: "De Bruyne", age: 33 }
+                    ],
+                },
```

Comment:
Can you please add a description comment before each scenario? Like, this one is probably "No stadium (it's optional)" or so

Reply:
Yep.